File: ui/src/lib/backend/persona.ts
/** Client for the persona folder — his fragments, editable. */

export interface PersonaFragment {
  /** Path relative to the persona folder, which is also its identity and its order. */
  name: string;
  /** "40-how-you-work.md" → "how you work". The numeric prefix is ordering, not a name. */
  title: string;
  enabled: boolean;
  /** The file, all of it. What a text editor would show you. */
  chars: number;
  /**
   * The part of the file that reaches the model — comments stripped, edges trimmed.
   *
   * The one that answers "what does this cost me", and the one that was missing. `chars` was
   * being shown under a heading about what every request carries, and on this repo's own persona
   * the two disagree by 31%: `35-how-you-spend-a-round.md` is 2,580 bytes of which 910 are his.
   * The rest is the note explaining why the fragment exists, which he never sees.
   */
  promptChars: number;
  body: string;
}

/** One fragment's contribution to the prompt, as the model receives it. */
export interface PersonaPart {
  name: string;
  text: string;
}

export interface PersonaState {
  folder: string;
  fragments: PersonaFragment[];
  /**
   * The same text as `merged`, still labelled with the file each piece came from.
   *
   * `merged` is literally `parts.map(p => p.text).join("\n\n")` on the server, so showing the
   * parts cannot show something he does not receive — which is the only reason it is safe to
   * render anything other than `merged` itself.
   */
  parts: PersonaPart[];
  /** What actually reaches the model: fragments joined, comments stripped. */
  merged: string;
  chars: number;
}

export async function fetchPersona(): Promise<PersonaState> {
  const response = await fetch("/api/persona");
  if (!response.ok) throw new Error(`/api/persona returned ${response.status}`);
  return (await response.json()) as PersonaState;
}

/**
 * Every write returns the fragment as it now is on disk, and callers get it.
 *
 * It used to return `Promise<void>` and drop the body, which cost more than a refetch. Enabling a
 * fragment is a *rename* on the server — the underscore prefix is the on-disk convention — so a
 * caller that wants to keep its selection has to know the new filename. With the response thrown
 * away the only way to know it was to re-derive the rename rule in TypeScript, and the copy was
 * wrong: it prefixed the first path segment rather than the filename, so disabling anything in a
 * subfolder predicted `_sub/10-x.md` where the server wrote `sub/_10-x.md`. The server already
 * knew the answer. Now it is allowed to say it.
 */
async function send(url: string, method: string, body?: unknown): Promise<PersonaFragment> {
  const response = await fetch(url, {
    method,
    headers: { "Content-Type": "application/json" },
    ...(body === undefined ? {} : { body: JSON.stringify(body) }),
  });
  const payload = (await response.json().catch(() => ({}))) as PersonaFragment & { error?: string };
  if (!response.ok) throw new Error(payload.error ?? `that didn't work (${response.status})`);
  return payload;
}

const path = (name: string) => `/api/persona/${name.split("/").map(encodeURIComponent).join("/")}`;

export const savePersonaFragment = (name: string, body: string) =>
  send(path(name), "PUT", { body });
export const addPersonaFragment = (name: string, body: string) =>
  send("/api/persona", "POST", { name, body });
export const setPersonaFragmentEnabled = (name: string, enabled: boolean) =>
  send(path(name), "PATCH", { enabled });
export const renamePersonaFragment = (name: string, next: string) =>
  send(path(name), "PATCH", { name: next });
/** The one that has nothing to return. */
export const deletePersonaFragment = async (name: string): Promise<void> => {
  await send(path(name), "DELETE");
};

File: server/kith/api/routes/persona.py
"""Who he is — the fragment files, editable.

The persona is a folder of markdown, and that stays true: these endpoints are file
operations, not a database behind a form. Someone who prefers a text editor can open the
folder and get the same result, which is the point of keeping it as files.
"""

from __future__ import annotations

from flask import jsonify, request

from kith.api.blueprint import api
from kith.services import persona


def _fail(exc: Exception, status: int = 400):
    return jsonify({"error": str(exc)}), status


@api.get("/persona")
@api.doc(
    summary="His persona, fragment by fragment",
    description=(
        "Every fragment including the disabled ones, in merge order, with its text. "
        "`merged` is what actually reaches the model — the fragments joined, comments "
        "stripped — so the size of the thing being edited is visible while editing it. "
        "`parts` is that same text still separated by the file it came from, because a "
        "reader who finds a sentence in the merge needs to know which fragment to open."
    ),
)
def get_persona():
    parts = persona.merged_parts()
    merged = "\n\n".join(part["text"] for part in parts)
    return jsonify(
        {
            "folder": str(persona.persona_dir()),
            "fragments": persona.fragments(),
            "parts": parts,
            "merged": merged,
            "chars": len(merged),
        }
    )


@api.post("/persona")
@api.doc(summary="Add a fragment", description="Refuses to overwrite an existing one.")
def create_fragment():
    payload = request.get_json(silent=True) or {}
    try:
        return jsonify(
            persona.create_fragment(str(payload.get("name") or ""), str(payload.get("body") or ""))
        )
    except persona.PersonaError as exc:
        return _fail(exc)


@api.put("/persona/<path:name>")
@api.doc(summary="Save a fragment")
def save_fragment(name: str):
    payload = request.get_json(silent=True) or {}
    try:
        return jsonify(persona.write_fragment(name, str(payload.get("body") or "")))
    except persona.PersonaError as exc:
        return _fail(exc)


@api.patch("/persona/<path:name>")
@api.doc(
    summary="Enable, disable, or rename a fragment",
    description=(
        "Both are renames underneath: disabling prefixes the filename with an underscore, "
        "which is the convention the loader reads, and the numeric prefix IS the merge "
        "order — so 'move this earlier' and 'rename this' are the same operation."
    ),
)
def patch_fragment(name: str):
    payload = request.get_json(silent=True) or {}
    try:
        if "enabled" in payload:
            return jsonify(persona.set_enabled(name, bool(payload["enabled"])))
        if payload.get("name"):
            return jsonify(persona.rename_fragment(name, str(payload["name"])))
    except persona.PersonaError as exc:
        return _fail(exc)
    return _fail(ValueError("nothing to change: send `enabled` or `name`"))


@api.delete("/persona/<path:name>")
@api.doc(summary="Remove a fragment", description="No undo — the interface asks first.")
def delete_fragment(name: str):
    try:
        persona.delete_fragment(name)
    except persona.PersonaError as exc:
        return _fail(exc, 404)
    return jsonify({"ok": True})
